safe_control_gym/controllers/pid/pid.py

```diff
@@ -259,10 +261,10 @@ def reset_before_run(self, obs=None, info=None, env=None):
         self.last_rpy = np.zeros(3)
         self.integral_rpy_e = np.zeros(3)
 
-        if env is None:
-            self.GRAVITY = self.g * self.env.OVERRIDDEN_QUAD_MASS
-        else:
-            self.GRAVITY = self.g * env.OVERRIDDEN_QUAD_MASS
+        # if env is None:
```

Comment:
Unsure if we should reset the model when the env is reset. I would say yes, since if you added randomization it should trigger whenever it is reset

Reply:
the current randomization only happens once when the ctrl is initialized. i think you are referring to randomizing the prior per epsiode? I'm not sure if this makes sense. since if I imagine a ctrl will learn based on a prior defined/randomized at initialization but fixed throughout training, then if we need to estimate the average performance, we launch multiple training of the controller and each time a training uses one randomized prior. Is there another use case you are thinking of?

Reply:
Ah i guess I was thinking in a more classical sense. For me, when ctrl.reset() is called, its a completely different experiment, so everything should be randomized again, but I see why it would be preferable to have a more explicit way to re-randomizing the prior. How do the inertial parameters get randomized? only once when the environment is initialized?

Reply:
oh i see, you mean doing prior randomization in ctrl.reset().  i thought you mean in ctrl.reset_before_run() since that's called with env.reset().

Reply:
Right yeah env.reset() makes more sense than env.reset_before_run(). Would there be a problem in reseting the prior there?

Reply:
the inertial prop can either be randomized externally and then instantiate env via `inertial_prop`, or it can instantiate env via `inertial_prop_rand_dict` to randomize internally (in the env, so each env.reset() will re-randomize). I use the former case to test robustness of ctrl against different inertial props and the latter case to do domain randomization training.

Reply:
Makes sense. if inertial prop is randomized on every reset, then so should the prior rt

Reply:
i believe u mean `Right yeah ctrl.reset() makes more sense than ctrl.reset_before_run().`?  

right, i think randomizing prior in `ctrl.reset()` also makes more sense, but i concern that some ctrl might need to use the prior to initialize things for training in `__init__`.  i guess we might somehow need to call `get_prior()` twice, one in `ctrl.__init__` and one in `ctrl.reset()` ?

Reply:
Yeah I think thats fine

Reply:
also on the env randomization of inertial prop, i think randomizing for each episode is only useful for domain randomization. i dont think other controllers need this feature.

Reply:
I am planning on using it when I run 100 tests, each with 10% randomization. For those tests I want it to be random on every singular experiment

Reply:
are you randomizing the `prior_prop` used for ctrl or the `inertail_prop` used for env simulation?

Reply:
just randomizing prior_prop. Keep inertial_prop constant and vary how wrong my model is in different ways.

Reply:
for every singular experiment i think you need prior randomization in `ctrl.reset()`, and if it's for `inertial_prop`, i think u need to randomize the inertial_prop then assign it to `inertial_prop` to the env in each experiment. Do I understand it correctly?

Reply:
Wait Im not sure. My understanding would be I would set `inertial_prop_randomization` to None or False or whatever, and set `prior_prop_randomization` to True or a dict or whatever and then whenever I call Experiment.reset() it will reset the env (not changing the inertial prop) and reset the controller (which will randomize the prior). Does that make sense?

Reply:
yes exactly. so now we are missing the prior randomization in `ctrl.reset()`

Reply:
yeah

Reply:
ok. thought about that a bit. since in that case, we can just call another `get_prior()` in `ctrl.reset()`. i guess if your controller needs it u can just do that in your `ctrl.reset()` function?  the current `get_prior()` will return a optionally randomized prior (from its config), so it will be up to each controller where to call it?

Reply:
Yeah that makes sense, so it may as well go into pid.reset rn rt

Reply:
so it will be up to the user's intension, if you need the prior to construct say a GP or NN in `ctrl.__init__`, you instantiate the prior there; but if you plan to reuse an existing ctrl multiple times (by calling `ctrl.reset()`), you can call `get_prior()` inside `ctrl.reset()`. Would this work?

Reply:
oh i see, you wanted it that way in PID, ok

Reply:
ty!

Reply:
ok. done. please check here [link](https://github.com/utiasDSL/safe-control-gym/pull/77/commits/888571da9ca0f0fe012306e8db8719b8c2e4cc29) and see if it does what you need?

Reply:
Yeah that looks great, thank you!

Reply:
ok. i will proceed to merging then?

Reply:
Yeah I think so, thank you!